RealFIFO: send DELETE events if an object has changed UID while disconnected from apiserver

munnerz opened the pull request:
#### What type of PR is this?

/kind bug

#### What this PR does / why we need it:

If a reflector/informer/store observes object A with UID 'A-1', and then:

1) the reflector disconnects from the apiserver
2) while disconnected, object A is deleted and recreated with the same name, now with UID 'A-2'
3) upon reconnect, RealFIFO will send a Replaced event for object A without sending a prior DELETE event

This matches behaviour in other FIFO implementations, however leads to correctness issues in the sequence of events, leading to an 'impossible' stream of events. In practice, this is mitigated if the consumer uses a 'level based' reconciliation approach and/or finalizers to manage the deletion of the  object, however some clients do against best practices take actions based on calls to `Add`/`Update`/`Delete` calls to event handlers.

This is especially problematic if a user is reflecting the state of one store into another concrete apiserver store, whereby the thing calling `Pop` would otherwise need to translate a `Replace` call into a 'delete or replace' by inspecting its own store once again, effectively, re-implementing the missing parts of FIFO (and possibly losing atomicity).

#### Which issue(s) this PR is related to:

I need to open an issue :)

#### Does this PR introduce a user-facing change?
```release-note
client-go: add InformerDeleteEventsByUID feature that ensures DELETE events are fired by informers when an object changes UID while the informer/reflector is disconnected from the apiserver
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

cc @deads2k @liggitt

munnerz committed:
RealFIFO: send DELETE events if an object has changed UID while disconnected from apiserver

liggitt commented:
Hmm… but isn't sending a synthetic delete event first incorrect in another direction, since we don't know *when* the object was deleted?

munnerz commented:
You mean when it was deleted relative to other objects that were also deleted? We already don't guarantee any ordering during a resync, but we *do* know that it happened *after* the current store state, and *before* the new state/now.

munnerz commented:
> I'm not sure this is a bug... the cache state is consistent with a real state that existed at a real point in time, which is what RealFIFO was trying to fix... I'd have to think about whether emitting a synthetic delete event when the uid changes on relist is always strictly better for observers.

It isn't though - while we are resyncing, we already dispatch synthetic deletes for items that are in the queue (or known in the store) that do not exist in the remote store. These are dispatched as regular 'cache.Deleted' events, prior to the stream of cache.Replaced events.

The states during this replay time are necessarily/by definition not mappable to absolute states of the store in the corresponding apiserver watched by the reflector. IMO additionally sending those deletes if the UID has _changed_ is no 'less correct' than our existing synthetic deletes.

I can/do/totally do share the sentiment around "is this *better* for the general case/all informers", especially given where we are up to in the ecosystem now/how far we have come without it.

That said, from the point of RealFIFO, even if this isn't exposed/used in informers, it seems like this leads to a more consistent state of the world than the alternatives. (for reference, I am using a reflector wired with a RealFIFO to then send Apply/Delete events to _another_ apiserver, hence running into the UID collisions - specifically, this *only happens if resources are re-created with the same name* - hence my feeling that this is somewhat of a bug....).

liggitt commented:
> IMO additionally sending those deletes if the UID has _changed_ is no 'less correct' than our existing synthetic deletes.

It seems like sending an `retErr := f.addToItems_locked(Replaced, false, obj)` entry is still the most correct thing

The thing handling the Replaced event has all the information it needs to detect if uid changed and decide to call the OnDelete/OnAdd handlers if it wants to, right? Something more like this?

```diff
diff --git a/staging/src/k8s.io/client-go/tools/cache/controller.go b/staging/src/k8s.io/client-go/tools/cache/controller.go
index 5f983b6b63a..8b4efc25888 100644
--- a/staging/src/k8s.io/client-go/tools/cache/controller.go
+++ b/staging/src/k8s.io/client-go/tools/cache/controller.go
@@ -563,7 +564,26 @@ func processDeltas(
 		obj := d.Object
 
 		switch d.Type {
-		case Sync, Replaced, Added, Updated:
+		case Replaced:
+			if old, exists, err := clientState.Get(obj); err == nil && exists {
+				if err := clientState.Update(obj); err != nil {
+					return err
+				}
+				oldUID := "" //...
+				newUID := "" //...
+				if oldUID == newUID {
+					handler.OnUpdate(old, obj)
+				} else {
+					handler.OnDelete(DeletedFinalStateUnknown{Obj: old})
+					handler.OnAdd(obj, isInInitialList)
+				}
+			} else {
+				if err := clientState.Add(obj); err != nil {
+					return err
+				}
+				handler.OnAdd(obj, isInInitialList)
+			}
+		case Sync, Added, Updated:
 			if old, exists, err := clientState.Get(obj); err == nil && exists {
 				if err := clientState.Update(obj); err != nil {
 					return err
```

lmktfy commented:
We could compromise and emit something like `InferredDelete`.

liggitt commented:
> We still send Replace (rather than Added) as this is 'replacing' empty in the store with the most up to date, ie it is not an Added event.
> 
> Process functions can still determine an object has been "magicked into existence at an indeterminate time" by checking if the object does or does not exist in the store.

Maybe I'm mistracing, but it looks like this PR will, for uid-changed objects,:
1. call addToItems_locked with Deleted / DeletedFinalStateUnknown 
2. call addToItems_locked with Replaced / current obj

As the controller is then processing those two deltas, it will:
1. handle the Deleted event first, remove the object from the store, and call handler.OnDelete
https://github.com/kubernetes/kubernetes/blob/4e8b192b66cc2a6952b8f1a5067e563c4019c276/staging/src/k8s.io/client-go/tools/cache/controller.go#L578-L582
2. handle the Replaced event, not find the object in the store, add the object to the store, and call handler.OnAdd
https://github.com/kubernetes/kubernetes/blob/4e8b192b66cc2a6952b8f1a5067e563c4019c276/staging/src/k8s.io/client-go/tools/cache/controller.go#L566-L577

That means between 1 and 2, the OnDelete handler function and any asynchronous reads from the store (lists / gets / etc) will not see the object in the store. It's not clear to me that exposing things that list from the store to states where the object is completely gone is better than what we have today.

An approach like https://github.com/kubernetes/kubernetes/pull/133565#issuecomment-3191420868 would still detect the uid change, would still emit OnDelete / OnAdd calls, but would make sure the object was never *gone* from the store.
